### PR TITLE
tailcfg, localapi: plumb device token to server

### PR DESCRIPTION
**Pull request by crawshaw**
questions in the form of a PR:

- I need to get this value out of the iOS app, not the networkextension, and localapi was handy. Is this a reasonable use of localAPI?
- should we send new stuff to control in Hostinfo? is there somewhere else this belongs? (it has the awkward property that on start the device token will be empty which I'm not a fan of, but otherwise looks like hostinfo stuff.)

### CODE REVIEW

**Comment by bradfitz** [comment]
Context? Link to bug?

**Comment by crawshaw** [comment]
https://github.com/tailscale/corp/issues/8940

File: ipn/localapi/localapi.go [hunk 1]
```diff
@@ -1200,6 +1201,24 @@ func (h *Handler) serveDial(w http.ResponseWriter, r *http.Request) {
 	<-errc
 }
 
+func (h *Handler) serveDeviceToken(w http.ResponseWriter, r *http.Request) {
+	if r.Method != "POST" {
+		http.Error(w, "unsupported method", http.StatusMethodNotAllowed)
+		return
+	}
```
Comment (bradfitz) [inline ipn/localapi/localapi.go:1208]: need permission checks here. See PermitWrite elsewhere in this file.

Is this only for setting the token? "set-device-token" and "serveSetDeviceToken"? Or will you add GET later so the frontend can decide whether it needs to do the init.

File: ipn/localapi/localapi.go [hunk 2]
```diff
@@ -1200,6 +1201,24 @@ func (h *Handler) serveDial(w http.ResponseWriter, r *http.Request) {
...
+	var params struct {
+		deviceToken string `json:"deviceToken"`
```
Comment (bradfitz) [inline ipn/localapi/localapi.go:1210]: capital DeviceToken

File: tailcfg/tailcfg.go [hunk 3]
```diff
@@ -525,6 +525,7 @@ type Hostinfo struct {
 	Desktop         opt.Bool       `json:",omitempty"` // if a desktop was detected on Linux
 	Package         string         `json:",omitempty"` // Tailscale package to disambiguate ("choco", "appstore", etc; "" for unknown)
 	DeviceModel     string         `json:",omitempty"` // mobile phone model ("Pixel 3a", "iPhone12,3")
+	DeviceToken     string         `json:",omitempty"` // macOS/iOS APNs device token for notifications
```
Comment (bradfitz) [inline tailcfg/tailcfg.go:528]: could be android too in the future

**Review by bradfitz — COMMENTED** [review]
Seems fine overall

File: ipn/localapi/localapi.go (hunk 2 above)
Comment (crawshaw) [reply ipn/localapi/localapi.go:1210]: just the field or also the json parameter name? we seem to have both kinds of parameter names?

Comment (bradfitz) [reply ipn/localapi/localapi.go]: You can't json.Unmarshal (due to `reflect` rules) into a field that's not exported. JSON parameter name doesn't matter.

But really this type shouldn't be defined here like this. It should be in a package where the client side can use it, even if the sole client currently in Swift that doesn't need it.

Comment (crawshaw) [reply ipn/localapi/localapi.go]: Yeah I changed it last minute after testing all the code.

Does the localapi have tests I should be extending for a new endpoint?

What package would the type go in?

Comment (crawshaw) [reply ipn/localapi/localapi.go]: (the question about the parameter name was a style question, I've no opinion other than I couldn't figure out what the standard was here.)

Comment (bradfitz) [reply ipn/localapi/localapi.go]:
```go
// Package apitype contains types for the Tailscale LocalAPI and control plane API.
package apitype
```

File: client/tailscale/apitype/apitype.go [hunk 4]
```diff
@@ -32,3 +32,9 @@ type WaitingFile struct {
 	Name string
 	Size int64
 }
+
+// SetDeviceToken is the body POSTed to /set-device-token.
```
Comment (bradfitz) [inline client/tailscale/apitype/apitype.go:36]: SetDeviceTokenRequest ?

mention LocalAPI in the docs too, as this package is a weird mix.

File: client/tailscale/apitype/apitype.go [hunk 5]
```diff
@@ -32,3 +32,9 @@ type WaitingFile struct {
...
+// It contains the iOS/macOS APNs device token (and any future Android equivalent).
```
Comment (bradfitz) [inline client/tailscale/apitype/apitype.go:37]: move this line down a line to be a doc on the field? "DeviceToken is ...."

File: hostinfo/hostinfo.go [hunk 6]
```diff
@@ -153,12 +154,16 @@ func GetEnvType() EnvType {
 }
 
 var (
+	deviceTokenAtomic atomic.Value // of string (base64)
```
Comment (bradfitz) [inline hostinfo/hostinfo.go:157]: what's this "base64" doing here? it's just a string, no?

base64 is a detail for another layer?

File: tailcfg/tailcfg.go [hunk 7]
```diff
@@ -525,6 +525,7 @@ type Hostinfo struct {
 	Desktop         opt.Bool       `json:",omitempty"` // if a desktop was detected on Linux
 	Package         string         `json:",omitempty"` // Tailscale package to disambiguate ("choco", "appstore", etc; "" for unknown)
 	DeviceModel     string         `json:",omitempty"` // mobile phone model ("Pixel 3a", "iPhone12,3")
+	DeviceToken     string         `json:",omitempty"` // macOS/iOS APNs device token for notifications (and maybe Android in the future)
```
Comment (bradfitz) [inline tailcfg/tailcfg.go:528]: should we call this NotifyToken? MobileNotifyToken?

**Review by bradfitz — APPROVED** [review]

File: tailcfg/tailcfg.go (hunk 7 above)
Comment (crawshaw) [reply tailcfg/tailcfg.go:528]: it's called DeviceToken in APNs. I'm not super keen on inventing a new name.

**Comment by crawshaw** [comment]
I wrote a test!

...which required fixing a bug in the host validation logic. 💀  PTAL

File: tailcfg/tailcfg.go (hunk 7 above)
Comment (bradfitz) [reply tailcfg/tailcfg.go:528]: Well I was thinking that if we use it for Android too it'll be wrong for one platform at least so we should make up a common term for both.

But it seems like Android also uses "Device Token" so sure! Good as is.

File: ipn/localapi/localapi.go [hunk 8]
```diff
@@ -202,7 +203,7 @@ func validHost(h string) bool {
 	if host == "localhost" {
 		return true
 	}
-	addr, err := netip.ParseAddr(h)
+	addr, err := netip.ParseAddr(host)
```
Comment (bradfitz) [inline ipn/localapi/localapi.go:206]: whoa, what? this should be its own commit

Comment (crawshaw) [reply ipn/localapi/localapi.go:206]: no good deed, etc etc

Comment (crawshaw) [reply ipn/localapi/localapi.go:206]: https://github.com/tailscale/tailscale/pull/7114

**Comment by crawshaw** [comment]
PTAL

File: tailcfg/tailcfg.go (hunk 7 above)
Comment (mihaip) [reply tailcfg/tailcfg.go:528]: Still think that having something about "push" or "notification" in the name would be useful, since it's next to a bunch of other generic fields (`DeviceModel`, `Package`) and it's not clear what kind of token we're talking about.

Comment (crawshaw) [reply tailcfg/tailcfg.go]: ok done!

File: ipn/localapi/localapi.go [hunk 9]
```diff
@@ -1198,6 +1199,26 @@ func (h *Handler) serveDial(w http.ResponseWriter, r *http.Request) {
...
+		return
+	}
+	var params apitype.SetPushDeviceTokenRequest
+	if err := json.NewDecoder(r.Body).Decode(&params); err != nil {
+		http.Error(w, "invalid JSON body", 400)
+		return
+	}
+	if params.PushDeviceToken != "" {
+		hostinfo.SetPushDeviceToken(params.PushDeviceToken)
+	}
```
Comment (bradfitz) [inline ipn/localapi/localapi.go:1218]: don't want to support any sort of clearing operation here?

File: tailcfg/tailcfg.go [hunk 10]
```diff
@@ -525,6 +525,7 @@ type Hostinfo struct {
 	Desktop         opt.Bool       `json:",omitempty"` // if a desktop was detected on Linux
 	Package         string         `json:",omitempty"` // Tailscale package to disambiguate ("choco", "appstore", etc; "" for unknown)
 	DeviceModel     string         `json:",omitempty"` // mobile phone model ("Pixel 3a", "iPhone12,3")
+	PushDeviceToken string         `json:",omitempty"` // macOS/iOS APNs device token for notifications (and maybe Android in the future)
```
Comment (bradfitz) [inline tailcfg/tailcfg.go:528]: s/maybe //

**Review by bradfitz — APPROVED** [review]

**Review by mihaip — COMMENTED** [review]
New name LGTM

Out of curiosity, do we have any kind of mechanism to force the controlclient to send an updated hostinfo when the token is set?

**Comment by bradfitz** [comment]
Yeah we do. This should be connected to that. Can look later when off phone.

**Comment by crawshaw** [comment]
I couldn't find an obvious thing for pushing a hostinfo update so I added one. PTAL?

File: ipn/ipnlocal/local.go [hunk 11]
```diff
@@ -1811,6 +1811,21 @@ func (b *LocalBackend) readPoller() {
 	}
 }
 
+// HostinfoUpdated is called to recompute the Hostinfo and send
+// the new version to the control server.
+func (b *LocalBackend) HostinfoUpdated() {
```
Comment (bradfitz) [inline ipn/ipnlocal/local.go:1816]: this name sounds like a accessor that returns a bool

make it sound imperative? `ResendHostinfoIfNeeded` maybe? Otherwise LGTM.

Wow, I forgot how horrible that `doSetHostinfoFilterServices` is. TODOs about in that call chain. (cc @andrew-d)

**Review by bradfitz — APPROVED** [review]

File: ipn/ipnlocal/local.go (hunk 11 above)
Comment (crawshaw) [reply ipn/ipnlocal/local.go]: done.

some of the hostinfo call chain looks like it wouldn't be too hard to clean up.
